services/horizon/internal/init.go

```
@@ -19,33 +19,70 @@ import (
 	"github.com/stellar/go/support/log"
 )
 
-func mustInitHorizonDB(app *App) {
-	session, err := db.Open("postgres", app.config.DatabaseURL)
+const maxIngestionDBConnections = 2
+
+func mustNewDBSession(databaseURL string, maxIdle, maxOpen int) *db.Session {
+	session, err := db.Open("postgres", databaseURL)
 	if err != nil {
 		log.Fatalf("cannot open Horizon DB: %v", err)
 	}
 
-	session.DB.SetMaxIdleConns(app.config.HorizonDBMaxIdleConnections)
-	session.DB.SetMaxOpenConns(app.config.HorizonDBMaxOpenConnections)
-	app.historyQ = &history.Q{session}
+	session.DB.SetMaxIdleConns(maxIdle)
+	session.DB.SetMaxOpenConns(maxOpen)
+	return session
 }
 
-func mustInitCoreDB(app *App) {
-	session, err := db.Open("postgres", app.config.StellarCoreDatabaseURL)
-	if err != nil {
-		log.Fatalf("cannot open Core DB: %v", err)
+func mustInitHorizonDB(app *App) {
+	maxIdle := app.config.HorizonDBMaxIdleConnections
+	maxOpen := app.config.HorizonDBMaxOpenConnections
+	// There are two connection pools, one for serving requests to horizon
+	// and another pool for ingestion. Ideally the total connections in both
+	// pools should be bounded by HorizonDBMaxOpenConnections. But, if the ingestion
+	// pool consumes a significant quota of HorizonDBMaxOpenConnections then we will
+	// allow a total of HorizonDBMaxOpenConnections + maxIngestionDBConnections connections.
```

I think we should just return an error (maybe at a command validation stage in `root.go`) when ingestion connections is equal/greater than max set by a user. The reason is that user may explicitly set the max number of open connections in Postgres config so these extra 2 connections won't be established (and will probably generate a bunch of errors in the log).